File: VB/HideNavigationItemsExample.Module/DatabaseUpdate/Updater.vb
Imports System
Imports System.Linq
Imports DevExpress.Xpo
Imports DevExpress.ExpressApp
Imports DevExpress.Data.Filtering
Imports DevExpress.ExpressApp.Xpo
Imports DevExpress.Persistent.Base
Imports DevExpress.ExpressApp.Updating
Imports DevExpress.Persistent.BaseImpl
Imports DevExpress.ExpressApp.Security
Imports HideNavigationItemsExample.Module.Security
Imports DevExpress.ExpressApp.Security.Strategy
Imports DevExpress.Persistent.BaseImpl.PermissionPolicy

Namespace HideNavigationItemsExample.Module.DatabaseUpdate
	Public Class Updater
		Inherits ModuleUpdater

		Public Sub New(ByVal objectSpace As IObjectSpace, ByVal currentDBVersion As Version)
			MyBase.New(objectSpace, currentDBVersion)
		End Sub
		Public Overrides Sub UpdateDatabaseAfterUpdateSchema()
			MyBase.UpdateDatabaseAfterUpdateSchema()
			Dim userRole As CustomSecurityRole = CreateUserRole()
			Dim administratorRole As CustomSecurityRole = CreateAdministratorRole()
			Dim admin As PermissionPolicyUser = ObjectSpace.FindObject(Of PermissionPolicyUser)(New BinaryOperator("UserName", "Admin"))
			If admin Is Nothing Then
				admin = ObjectSpace.CreateObject(Of PermissionPolicyUser)()
				admin.UserName = "Admin"
				admin.IsActive = True
				admin.SetPassword("")
				admin.Roles.Add(administratorRole)
				admin.Save()
			End If
			Dim user As PermissionPolicyUser = ObjectSpace.FindObject(Of PermissionPolicyUser)(New BinaryOperator("UserName", "User"))
			If user Is Nothing Then
				user = ObjectSpace.CreateObject(Of PermissionPolicyUser)()
				user.UserName = "User"
				user.IsActive = True
				user.SetPassword("")
				user.Roles.Add(userRole)
				user.Save()
			End If
		End Sub
		Private Function CreateAdministratorRole() As CustomSecurityRole
			Dim administratorRole As CustomSecurityRole = ObjectSpace.FindObject(Of CustomSecurityRole)(New BinaryOperator("Name", SecurityStrategyComplex.AdministratorRoleName))
			If administratorRole Is Nothing Then
				administratorRole = ObjectSpace.CreateObject(Of CustomSecurityRole)()
				administratorRole.Name = SecurityStrategyComplex.AdministratorRoleName
				administratorRole.IsAdministrative = True
			End If
			Return administratorRole
		End Function
		Private Function CreateUserRole() As CustomSecurityRole
			Dim userRole As CustomSecurityRole = ObjectSpace.FindObject(Of CustomSecurityRole)(New BinaryOperator("Name", "Users"))
			If userRole Is Nothing Then
				userRole = ObjectSpace.CreateObject(Of CustomSecurityRole)()
				userRole.Name = "Users"
				userRole.SetTypePermission(Of Person)(SecurityOperations.FullAccess, SecurityPermissionState.Allow)
				userRole.SetTypePermission(Of PhoneNumber)(SecurityOperations.FullAccess, SecurityPermissionState.Allow)
				userRole.SetTypePermission(Of Address)(SecurityOperations.FullAccess, SecurityPermissionState.Allow)
				userRole.SetTypePermission(Of Country)(SecurityOperations.FullAccess, SecurityPermissionState.Allow)
				userRole.AddObjectPermission(Of SecuritySystemUser)(SecurityOperations.ReadOnlyAccess, "[Oid] = CurrentUserId()", SecurityPermissionState.Allow)
				userRole.HiddenNavigationItems = "AddressGroup, Person_Varied"
			End If
			Return userRole
		End Function
	End Class
End Namespace

File: CS/HideNavigationItemsExample.Module/DatabaseUpdate/Updater.cs
using System;
using System.Linq;
using DevExpress.Xpo;
using DevExpress.ExpressApp;
using DevExpress.Data.Filtering;
using DevExpress.ExpressApp.Xpo;
using DevExpress.Persistent.Base;
using DevExpress.ExpressApp.Updating;
using DevExpress.Persistent.BaseImpl;
using DevExpress.ExpressApp.Security;
using HideNavigationItemsExample.Module.Security;
using DevExpress.ExpressApp.Security.Strategy;

namespace HideNavigationItemsExample.Module.DatabaseUpdate {
    public class Updater : ModuleUpdater {
        public Updater(IObjectSpace objectSpace, Version currentDBVersion) :
            base(objectSpace, currentDBVersion) {
        }
        public override void UpdateDatabaseAfterUpdateSchema() {
            base.UpdateDatabaseAfterUpdateSchema();
            CustomSecurityRole userRole = CreateUserRole();
            CustomSecurityRole administratorRole = CreateAdministratorRole();
            SecuritySystemUser admin = ObjectSpace.FindObject<SecuritySystemUser>(new BinaryOperator("UserName", "Sam"));
            if (admin == null) {
                admin = ObjectSpace.CreateObject<SecuritySystemUser>();
                admin.UserName = "Sam";
                admin.IsActive = true;
                admin.SetPassword("");
                admin.Roles.Add(administratorRole);
                admin.Save();
            }
            SecuritySystemUser user = ObjectSpace.FindObject<SecuritySystemUser>(new BinaryOperator("UserName", "John"));
            if (user == null) {
                user = ObjectSpace.CreateObject<SecuritySystemUser>();
                user.UserName = "John";
                user.IsActive = true;
                user.SetPassword("");
                user.Roles.Add(userRole);
                user.Save();
            }
        }
        private CustomSecurityRole CreateAdministratorRole() {
            CustomSecurityRole administratorRole = ObjectSpace.FindObject<CustomSecurityRole>(
                new BinaryOperator("Name", SecurityStrategyComplex.AdministratorRoleName));
            if (administratorRole == null) {
                administratorRole = ObjectSpace.CreateObject<CustomSecurityRole>();
                administratorRole.Name = SecurityStrategyComplex.AdministratorRoleName;
                administratorRole.IsAdministrative = true;
            }
            return administratorRole;
        }
        private CustomSecurityRole CreateUserRole() {
            CustomSecurityRole userRole = ObjectSpace.FindObject<CustomSecurityRole>(
                new BinaryOperator("Name", "Users"));
            if (userRole == null) {
                userRole = ObjectSpace.CreateObject<CustomSecurityRole>();
                userRole.Name = "Users";
                userRole.SetTypePermissions<Person>(SecurityOperations.FullAccess, SecuritySystemModifier.Allow);
                userRole.SetTypePermissions<PhoneNumber>(SecurityOperations.FullAccess, SecuritySystemModifier.Allow);
                userRole.SetTypePermissions<Address>(SecurityOperations.FullAccess, SecuritySystemModifier.Allow);
                userRole.SetTypePermissions<Country>(SecurityOperations.FullAccess, SecuritySystemModifier.Allow);
                userRole.AddObjectAccessPermission<SecuritySystemUser>("[Oid] = CurrentUserId()", SecurityOperations.ReadOnlyAccess);
                userRole.HiddenNavigationItems = "AddressGroup, Person_Varied";
            }
            return userRole;
        }
    }
}
